independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java

```diff
@@ -74,6 +74,12 @@
 
     @Override
     public void fire(T event) {
+        // If the runtime type of the event object contains an unresolved type variable, the container must throw an IllegalArgumentException.
+        Type eventType = getEventType(event.getClass());
+        if (Types.containsTypeVariable(eventType)) {
+            throw new IllegalArgumentException(
+                    "CDI event payload cannot contain unresolved type variable; found type: " + eventType);
+        }
         Objects.requireNonNull(event, "Event cannot be null");
```

Comment:
This line should be above the newly added lines, not below :-) Alternatively, it can be removed, as `event.getClass()` will throw a NPE on its own.

Reply:
Good point, thanks.
Fixed and rebased onto `main`